src/analysis/seminorm.lean

```diff
@@ -0,0 +1,219 @@
+/-
+Copyright (c) 2019 Jean Lo. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jean Lo
+-/
+
+import algebra.pointwise
+import analysis.normed_space.basic
+
+/-!
+# Seminorms and local convexity
+
+This file introduces the following notions, defined for a vector space
+over a normed field:
+
+- the subset properties of being `absorbent` and `balanced`,
+
+- a `seminorm`, a function to the reals that is positive-semidefinite,
+  absolutely homogeneous, and subadditive.
+
+We prove related properties.
+
+## TODO
+
+Define and show equivalence of two notions of local convexity
+for a t.v.s. over ℝ or ℂ: that it has a local base of balanced convex
+absorbent sets, and that it carries the initial topology induced by a
+family of seminorms.
+
+## References
+* [H. H. Schaefer, *Topological Vector Spaces*][schaefer1966]
+-/
+
+-- subset properties : absorbent and balanced sets in a vector space
+-- over a nondiscrete normed field
+section
+
+variables
+(𝕜 : Type*) [nondiscrete_normed_field 𝕜]
+{E : Type*} [add_comm_group E] [vector_space 𝕜 E]
+
+open set normed_field
+
+/-- A set `A` absorbs another set `B` if `B` is contained in scaling
+`A` by elements of sufficiently large norms. -/
+def absorbs (A B : set E) := ∃ r > 0, ∀ a : 𝕜, r ≤ ∥a∥ → B ⊆ a • A
+
+/-- A set is absorbent if it absorbs every singleton. -/
+def absorbent (A : set E) := ∀ x, ∃ r > 0, ∀ a : 𝕜, r ≤ ∥a∥ → x ∈ a • A
+
+/-- A set `A` is balanced if `a • A` is contained in `A` whenever `a`
+has norm no greater than one. -/
+def balanced (A : set E) := ∀ a : 𝕜, ∥a∥ ≤ 1 → a • A ⊆ A
+
+
+variables {𝕜} (a : 𝕜) {A : set E}
+
+/-- A balanced set absorbs itself. -/
+lemma absorbs_self_of_balanced (hA : balanced 𝕜 A) : absorbs 𝕜 A A :=
```

Comment:
If you rename the lemma to `balanced.absorbs_self`, then it will be possible to write `hA.absorbs_self`, where `hA : balanced 𝕜 A`.

```suggestion
lemma balanced.absorbs_self (hA : balanced 𝕜 A) : absorbs 𝕜 A A :=
```

Reply:
This, and all subsequent edits, have been effected as of 729e53ded4e5cd5421f9093e66f936e5becd3615 . Resolving conversations.